src/Compilers/Test/Core/Platform/Desktop/Extensions.cs

```diff
@@ -30,27 +31,27 @@ public static void AddArray<T>(this SerializationInfo info, string name, Immutab
             // we will copy the content into an array and serialize the copy
             // we could serialize element-wise, but that would require serializing
             // name and type for every serialized element which seems worse than creating a copy.
-            info.AddValue(name, value.IsDefault ? null : value.ToArray(), typeof(T[]));
+            info.AddValue(name, value.IsDefault ? null : ImmutableCollectionsMarshal.AsArray(value), typeof(T[]));
         }
 
         public static ImmutableArray<T> GetArray<T>(this SerializationInfo info, string name) where T : class
         {
-            var arr = (T[])info.GetValue(name, typeof(T[]));
-            return ImmutableArray.Create<T>(arr);
+            var array = (T[])info.GetValue(name, typeof(T[]));
+            return ImmutableCollectionsMarshal.AsImmutableArray(array);
         }
 
         public static void AddByteArray(this SerializationInfo info, string name, ImmutableArray<byte> value)
         {
             // we will copy the content into an array and serialize the copy
             // we could serialize element-wise, but that would require serializing
             // name and type for every serialized element which seems worse than creating a copy.
-            info.AddValue(name, value.IsDefault ? null : value.ToArray(), typeof(byte[]));
+            info.AddValue(name, value.IsDefault ? null : ImmutableCollectionsMarshal.AsArray(value), typeof(byte[]));
```

Comment:
Here as well

Reply:
Think those comments are somewhat wrong even in the before case. Let me clean that up.